docs/design-documents/hal/0001-i2c-overhaul.md

```diff
@@ -0,0 +1,527 @@
+
+  The length parameter cannot be signed.
+
+- **Change** the `data` parameter in `i2c_write` and `i2c_read` functions from `char*` to `uint8_t*`.
+
+### Slave API changes
+
+The main changes involve removing the slave specific read/write functions and rolling them into the normal read/write functions, removing most of the slave configuration which can be handled at construction by the `init` function.
+
+- **Remove** the `i2c_slave_mode` function, add an `is_slave` parameter to the `i2c_init` function.
+
+  The decision to initialise the peripheral in master or slave mode should be decided at construction time. This simplifies the API as it removes the need for two separate functions to initialise slave mode `i2c_slave_mode` and `i2c_slave_address` .
+
+- **Remove** the I2C slave specific transfer functions: `i2c_slave_read`, `i2c_slave_write`.
+
+  These functions are superfluous and can be rolled into the existing `i2c_read` and `i2c_write` functions. The `transfer` function will execute the slave read/write based on the current configuration of the peripheral.
+
+- **Change** `i2c_slave_receive` to `i2c_slave_status` and the return type from an integer to an enumeration.
+
+  The function returns which mode the peripheral is currently operating in as an integer, this is better expressed as an enumeration.
+
+### Async API changes
+
+- **Remove** the `DMAUsage` argument for asynchronous transfers.
+
+  Currently the `DMAUsage` argument of the `i2c_transfer_asynch` function is unimplemented, the argument is unused by all `I2C` implementations. There is no real demand for it so can be removed from the API.
+
+- **Change** the `stop` parameter from the `i2c_transfer_async` from an `uint32_t` to a `bool`.
+
+  The stop parameter indicates whether or not the function should send a `STOP` command after the transfer has complete, there is no reason at all for this to be a `uint32_t`.
+
+- **Change** the return type of `i2c_transfer_asynch` to indicate whether or not a transfer has been scheduled or not.
+
+  The function now returns a `bool` indicate if a transfer was scheduled or not, which can occur if the peripheral is already busy.
+
+- **Remove** the `i2c_irq_handler_asynch` function from the API.
+
+  The event is now passed as an argument to the callback this method is no longer required.
+
+- **Remove** the `event` parameter from the `i2c_transfer_async` function.
+
+  The callback will now be invoked on any event with the event as an argument.
+
+- **Remove** the `i2c_active` function from the API.
+
+  The the async callback is now always invoked on async operation  termination (unless cancelled), this status can be tracked from driver layer without any HAL request.
+
+### The new API
+
+```c++
+/** Error codes */
+enum {
+    I2C_ERROR_NO_SLAVE = -1,
+    I2C_ERROR_BUS_BUSY = -2,
+    I2C_ERROR_TIMEOUT  = -3,
+    I2C_ERROR_ARBITRATION_LOST = -4
+};
+
+typedef struct {
```

Comment:
Is it worth packing this struct?

Reply:
I'm not sure. Will save only `3B` per instance. An further savings will depend on how this struct will be utilized in reworked I2C drivers layer.